### PR TITLE
#556 Initial Cmake Module definition.

### PR DESCRIPTION
**Pull request by phelter**
Support for CMake and compilation with strictest definitions for GNU/Clang.

Description
-----------
Added support for CMake compile of the source code.
- Excludes some Ports at the moment - due to lack of testing infrastructure.
- Excludes integrating/changing unit-tests and compiling:
  - `test/Coverity`
  - `test/cbmc`
  - `test/unit-tests`
  - Reason is these are projects of their own and do not know the interworkings of these - could integrate later.
- Includes build of the Various Build-configurations for testing - including a CUSTOM for externally defined configuration.
  - `DEFAULT_CONF`, `ENABLE_ALL`, `DISABLE_ALL`

Test Steps
-----------
- Updated CI - but tested locally with:
  - Ubuntu 18.04 with Clang 13 - ran all 3 build configuration tests
  - Ubuntu 18.04 with GNU 7.5 - ran all 3 build configuration tests

Related Issue
-----------
- #556 - Support for CMake.
- Note there are several Clang warnings that should be addressed in various files.  These will be fixed in subsequent PR's.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by AniruddhaKanhere** [comment]
@phelter, thank you for raising the issue and creating this PR!
We are looking into it and will get back to you with changes (if any).

Thanks for your patience!

**Comment by phelter** [comment]
@AniruddhaKanhere - Thanks for reviewing.  I hope it is along the lines of how you want `FreeRTOS-Plus-TCP` to be integrated into projects.  Feel free to ask why certain things are done a certain way as I know CMake isn't well used in the embedded firmware realm too much and is an unknown beast.

It would be great if some of the `ports` could be done by others following the same CMakeLists.txt format/guidelines as the other ports and then I can review those as needed.

**Comment by htibosch** [comment]
> CMake isn't well used in the embedded firmware
> realm too much and is an unknown beast

That is probably why I know nothing about how to use cmake. I'm a GNU make user.

Like I just commented to your [PR #35](https://github.com/FreeRTOS/Lab-Project-FreeRTOS-FAT/pull/35), would you mind to write a few lines that explain how things can be tested?

And can you also update this PR with the main branch?

Also it needs formatting: you can start that to write a 1-line comment with the contents:
`/bot run uncrustify`
that should start reformatting. Don't forget to pull to your local branch after doing that.
Thanks

**Comment by phelter** [comment]
@htibosch 

- For how things are tested:
Please see the changes to the `ci.yml` file here: https://github.com/phelter/FreeRTOS-Plus-TCP/blob/cf5aa7ff657ba650a6e62f30aa45a58a60173397/.github/workflows/ci.yml#L110

The cmake file now not only compiles just for these tests, but also provides the static libraries for others to include in their builds.

The user can also use these test builds to confirm that their specific `FreeRTOSConfig.h` and `FreeRTOSIPConfig.h` are also valid for the current versions of the Kernel and Plus-TCP components.

If you would rather just look at the results the last runs are here:
https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/actions/runs/3200788444/jobs/5268962000

- And can you also update this PR with the main branch?
Done

- Formatting
Done as mentioned. `/bot run uncrustify`

**Comment by phelter** [comment]
@AniruddhaKanhere & @htibosch

I believe I've provided all of the requested info about this PR.  Would it be possible to add a few eyes on this PR in the near future?  

Thanks

File: README.md
Comment (archigup) [inline README.md:34]:
```suggestion
## To consume FreeRTOS+TCP
```

File: .github/workflows/ci.yml
Comment (archigup) [inline .github/workflows/ci.yml:116]:
```suggestion
```

File: tools/CMakeLists.txt
Comment (archigup) [inline tools/CMakeLists.txt:14]: Which warnings are these? Can the includes by fixed instead?

File: CMakeLists.txt
Comment (archigup) [inline CMakeLists.txt:166]: The kernel should compile with C89

File: source/portable/CMakeLists.txt
Comment (archigup) [inline source/portable/CMakeLists.txt:177]: Could you remove the commented out stuff?

File: tools/CMakeLists.txt
Comment (phelter) [reply tools/CMakeLists.txt:14]: Yes - the plan here is to add the necessary CMake compilation with minimal changes to the C files themselves and Identify all of the compilation warnings first.
One of them that was identified is `-Wpadded (clang)` - this means that there is additional padding on particular structs etc.  which in most cases is okay, but points to possible improvements for better layout in memory and/or alignment.

A follow on PR for actually removing these warnings will happen after this PR is merged.  I didn't want to have too much in each of the PR's.  What is currently there has been used/consumed for the longest time so assuming none of these are major issues.

File: CMakeLists.txt
Comment (phelter) [reply CMakeLists.txt:166]: @archigup - IMO, that ship has sailed.  Unless there is a significant amount of work to remove all of the code that is using C99 constructs and operation is done in both the FreeRTOS-Kernel and FreeRTOS-Plus-TCP this is not viable at the moment.

One example is `declaration after execution` - i.e. declaring variables midway in a procedure.  There are many cases.

Is there a particular compiler that doesn't support C99 at this point in time? - For those users who wish to continue compiling with a compiler with no C99 support, I believe if they are not likely to update their compiler, what are the chances they will consume a new version of FreeRTOS-Plus-TCP as well?  We are talking about 20+ yr old compilers here.

**Comment by phelter** [comment]
@archigup , @AniruddhaKanhere and @htibosch .

I've updated the PR and have fixed all of the requested issues identified.

Outstanding item: `C99` vs` C8`9 - please confirm you do or don't want this change at this moment. Note that the current FreeRTOS-Plus-TCP and FreeRTOS-Kernel both use `C99 ISO` constructs and would have to be re-worked to support `C89` without violating the ISO.

**Comment by AniruddhaKanhere** [comment]
Hello @phelter,

Thank you for making the changes that we suggested!

We are discussing the changes that you have proposed here (and in FreeRTOS-Kernel repository). We don't want to rush accepting your PR without a discussion just to later overwrite your changes. To that end, we are trying to figure out a better way to make the changes that you have proposed.

It is taking time - as all good things do :)

I shall get back to you in a bit once we have figured out the best way to make certain changes and accept your PR (and also the one in FreeRTOS-kernel repository).

@paulbartell had some suggestions which he will add to the PR comments.

Thanks for your patience @phelter and for your efforts with your PRs,
Aniruddha

File: CMakeLists.txt
Comment (paulbartell) [inline CMakeLists.txt:17]: why not?

File: CMakeLists.txt
Comment (paulbartell) [inline CMakeLists.txt:32]: Could each buffer allocation be exposed as a separate static lib target?

File: CMakeLists.txt
Comment (paulbartell) [inline CMakeLists.txt:54]: Could each of these be a separate static lib target?

File: CMakeLists.txt
Comment (paulbartell) [reply CMakeLists.txt:166]: @phelter : We have had a number of internal discussions on this topic and our current policy is ANSI C89 / ISO C90 compliance with a limited number of exceptions (stdbool and stdint mainly). I don't expect this policy to change in the near future until we can be 100% certain there are no compatibility breaks.

If there are specific C89/C90 compliance problems, they should be either:
1) Considered a bug <OR>
2) Added to the list of exceptions for that particular repository in the repository's README.md and masked for common compiler in the cmake or source files.

Note, this mainly applies to library code. I'ts a bit more flexible with demos or other examples that target a particular platform. In the case of FreeRTOS+TCP, the portable directory is excluded from this policy.

File: tools/CMakeLists.txt
Comment (paulbartell) [inline tools/CMakeLists.txt:3]: Could this been included in the default FreeRTOS+TCP target?

File: CMakeLists.txt
Comment (phelter) [reply CMakeLists.txt:166]: Ok - thanks - will assume all use of C99 as a bug then and identify them in the warnings for now - and fix them in a follow-on commit.

File: CMakeLists.txt
Comment (phelter) [reply CMakeLists.txt:17]: Since it is in other `CMakeList.txt` files in this current directory, I added the same restriction here.

See other examples: https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/search?q=if%28+%24%7BPROJECT_SOURCE_DIR%7D+STREQUAL+%24%7BPROJECT_BINARY_DIR%7D+%29

File: CMakeLists.txt
Comment (phelter) [reply CMakeLists.txt:32]: I treated this similar to how the HEAP selection is performed in FreeRTOS-Kernel.  

Option 1: Each BufferAllocation be exposed as separate static lib target:
-  +ves Always compiled and available.
-  -ves Compiling code that the user does not use.  Additional static libraries generated.
Option 2: port specific - it is not port specific feature - a port may limit the BufferAllocation used, but they are independent configurations.

Will leave as is, unless a specific change is requested.

File: CMakeLists.txt
Comment (phelter) [reply CMakeLists.txt:54]: Yes - it makes more sense to separate these out as independent `CMakeLists.txt` in each of the respective portable/* directories, but I followed the way the FreeRTOS-Kernel does it instead.

That way the only selection (single case/if-else selection) is in the `portable/CMakeLists.txt` - as to which `add_subdirectory()` to use.

File: tools/CMakeLists.txt
Comment (phelter) [reply tools/CMakeLists.txt:3]: Yes that `could` be done - if that is the case then the tools/* code `should` be placed in the same directory as the freertos_plus_tcp code.

Typically a separate directory means a separate library.  Having multiple libraries here is not a big deal.  The up-side to leaving as-is in the tcp_utilities directory allows one to take this and combine this static library in their testing area so they have the same code monitoring the counts inside and outside the `Unit under test`.  Otherwise one would need to expose a private set of API's in both.

Assuming leaving here as-is - for reasons provided.

File: CMakeLists.txt
Comment (phelter) [reply CMakeLists.txt]: FYI:

I've fixed all the C90 - ISO violations in the code and have separated out the cmakes so they are per Network IF port now instead of all one big lump with multiple `case` statements in the one.

I do not intend to test all of the other ports but have tested the linux and Zynq ports and have confirmed they work.

File: CMakeLists.txt
Comment (phelter) [reply CMakeLists.txt:54]: Yes - independent CMakeLists.txt now - which makes adding and maintaining the ports much simpler as they are now contained with their own CMakeLists.txt file.

File: CMakeLists.txt
Comment (phelter) [reply CMakeLists.txt]: @paulbartell - One other thing to note.  If you had example builds of each and every compiler that FreeRTOS-Plus-TCP (and FreeRTOS-Kernel) intends to support, then you could use those as a deciding factor as to whether you could increase to C99 or higher (eg C11).  The example builds could be easily provided in a CI/CD build by:
* Creating a docker image (eg. Linux Ubuntu) per compiler supported - along with installed BSP of each of the boards that is identified as supported.
* Provide a toolchain file per compiler
* Provide a toolchain file per board (optional) that includes the specific compile options for that board and includes the compiler toolchain (Dont-Repeat-Yourself DRY).  This is optional because if you only want to confirm the compiler works then you really only need one example and all others should be the same with the exception of the portable code that is compiled.  To reduce the amount of time checking all the different boards you could compile everything with one board BSP and then just compile the `freertos_plus_tcp_network_if` target for each of the other boards supported with that same compiler.
* Add the necessary build as a cross-compilation per compiler and board in the CI/CD flow - probably only necessary for release builds.

Then there is a guarantee that if you change something like the Required Compiler ISO - you know it won't break someone's build.

**Comment by phelter** [comment]
@paulbartell , @AniruddhaKanhere , @archigup :

Any progress on this? - would like to not rely on my forks of FreeRTOS/* for projects that I am working on.

File: source/portable/NetworkInterface/STM32Fxx/CMakeLists.txt
Comment (AniruddhaKanhere) [inline source/portable/NetworkInterface/STM32Fxx/CMakeLists.txt:15]:
```suggestion
    stm32fxx_hal_eth.h
```

**Review by AniruddhaKanhere — CHANGES_REQUESTED** [review]
Just a small NIT pick

File: tools/CMakeLists.txt
Comment (phelter) [reply tools/CMakeLists.txt:14]: If you want to see the warnings/changes - they are here:

* https://github.com/phelter/FreeRTOS-Plus-TCP/compare/feature/cmake-support...feature/fixing-clang-gnu-compiler-warnings

The big sets are:
* misuse of doxygen comments or out of date parameter names 
* uninitialized variables at the beginning - which have cases where they may be used before loaded/set - Undefined Behaviour.
* invalid use of unsigned vs. signed - missing casts.

Or you can compile as mentioned with clang-13 or 14 or above and check the warning lists yourselves.

**Comment by phelter** [comment]
@paulbartell , @AniruddhaKanhere , @archigup :

Just a FYI,

This is the last post and update you will get from me for these changes.

I am no longer going to maintain this.  Please feel free to add or abandon these changes.

I am moving my efforts on the opensource community away from FreeRTOS as the turnaround times for changes and PR's is in excess of 5 months and it looks as though there is little effort in maintaining this code base.

I am assuming Amazon is winding down it's support for FreeRTOS since there has been a shift/migration to [Zephyr](https://zephyrproject.org/) .

**Comment by AniruddhaKanhere** [comment]
Hello @phelter,

> I am moving my efforts on the opensource community away from FreeRTOS as the turnaround times for changes and PR's is in excess of 5 months

I am sorry to hear that and sincerely apologize for the delay. We did not use `CMake` in +TCP and the FreeRTOS-Kernel before and had many discussions about which way is the best to avoid rework and make the repository easier to use for our users. This led to all the discussions and comments we had on your PR.
Regardless, I share your view that the turnaround time on this PR was quite large. To fix that and to ensure that this is not repeated with any other PR, we are working on a mechanism which should help us focus on pending PRs amongst other things.

> and it looks as though there is little effort in maintaining this code base.

This might not be completely accurate. We were focused on adding and refactoring the changes in the IPv6 branch. IPv6 functionality is quite sought after and thus we were prioritizing that.

> I am assuming Amazon is winding down it's support for FreeRTOS

This is also not accurate. If anything, we are adding more and more to the FreeRTOS ecosystem.

Sincerely,
Aniruddha

**Review by AniruddhaKanhere — APPROVED** [review]
Thank you for this PR @phelter!

I am approving this PR. Once we have a second approval, we can merge it mainline

Thanks

**Comment by phelter** [comment]
Thanks @AniruddhaKanhere for the review.

Please note I believe this repo has similar issues to the ones mentioned in this issue: 
https://github.com/aws/ota-for-aws-iot-embedded-sdk/issues/465

And this was mentioned in this issue here - https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/issues/570#issuecomment-1312560393

It looks as though there is a fundamental misunderstanding in the requirements for making a `module and it's API` task/thread safe and robust in the various aws opensource projects.  I'd suggest adding a similar issue to this repo as well regarding a `State Access Audit`.

And to ensure these type of issues don't pop up with future modifications, consider incorporating both Address and Thread Sanitizer runtime unit tests. Where the unit tests stimulates the module as a `black/gray-box` (with all of the appropriate various threads and timers running in the typical threading organization).  These type of unit-tests can then also act as confirmation that the examples you provide are operational as they are more like example code rather than low-level unit-tests.

Why do I believe this is an issue in this repo?

I've been seeing various issues with FreeRTOS-Plus-TCP hanging when using the FTP server functionality.  It passes when it is compiled Debug mode, but when compiled for Release mode, the pings stop processing.  In my experience, these types of bugs are typically seen when there are issues with State Access between threads/tasks (or between ISR and thread contexts).  There were too many potential locations to list out for this particular repo and I was unable to convince @htibosch of the need to do this level of scrutiny without a specific error case that explicitly exposed this type of failure.

**Review by amazonKamath — APPROVED** [review]

**Comment by cobusve** [comment]
@phelter for what it is worth I believe you are right about the State Access Audit. We have already done a lot of work on this and will continue with doing that. If you have more suggestions I would love to hear them.

File: tools/CMakeLists.txt
Comment (amazonKamath) [reply tools/CMakeLists.txt:14]: Thanks @phelter.  CC: @kstribrnAmzn, @shubnil

**Pull request merged**
